### ISSUE TITLE
Reading BAM Records into a FragmentStore in batches

### ISSUE THREAD

**Issue by martinjvickers** [seqan/seqan#2121]
How do I go about reading BAM Records into a FragmentStore in batches?

I'm trying to work with the reads at specific bases within the TAIR10 genome. So first I created a FragmentStoreConfig to deal with Iupac bases and now I'd like to extract regions out, iterate over the genome etc. without using all the RAM.

```
	// Create a custom fragment store which accepts Iupac characters
	struct NewStoreConfig: public FragmentStoreConfig<>
	{
		typedef String<Iupac> TReadSeq;
		typedef String<Iupac> TContigSeq;
		typedef Owner<>	TReadSeqStoreSpec;
		typedef Owner<> TReadNameStoreSpec;
	};

	typedef FragmentStore<void, NewStoreConfig> TNewFragStore;
	TNewFragStore store;

	// load contigs into the store
	loadContigs(store, toCString(options.referenceFileName));

        // load reads into the store
        BamFileIn file(toCString(options.inputFileName));
        readRecords(store, file);

        // locks the machine :-(
```

A small BAM file (1.3GB - 28million reads) locks up my 32GB RAM workstation just attempting to read in. 

Is there a better way of doing this?

**Comment by smehringer** [comment]
Hi @martinjvickers,
we are very sorry for answering you so late.
Unfortunately the `FragementStore` does not support reading a bam file in batches and it would require some more work to include this feature. Is this issue still relevant for you?

A naive solution would be to either (1) you read in the records using a SeqFileIn object (which allows a `numRec` parameter) and then assign the batch to the fragment store or (2) to split/downsample your bam file beforehand if you are only interested in specific regions anyway.

**Comment by rrahn** [comment]
@smehringer do you think adding a batch wise read function can be done for 2.4?
Not sure how complicated the code base of the fragment store is.

**Comment by smehringer** [comment]
I'm not sure.. there seem to be a lot of things happening before the actual records are read and it will take time until one can be sure that adjustments are safe. Do you think it is important enough, since the author of this issue is also not responding?

**Comment by rrahn** [comment]
No, if it seems to complicate (I heard stories) than we leave it for now.
@martinjvickers please reopen if you feel that your problem is not resolved.

**Comment by martinjvickers** [comment]
Hi @smehringer and @rrahn, thanks for the response. I've not resolved this so it's still outstanding but I appreciate that the solution may not be a read in batch function or even using the fragment store.

@smehringer solution 2) that you've mentioned is something I would do if I simply needed to process some data and noone would use however as this might be something to release then I wouldn't want the instructions for use to be "split your BAM into X chunks of size XXXMB". (Out of curiosity, what is the intended input size and use-case for the FragmentStore, if it's using >32GB RAM for a 1.2GB input file it must be small BAM files?).

Solution 1) sounds like it might be what I need and was how I assumed originally it would work. I'm not sure how I would assign this though. Simply by adding the `numRec` parameter to the above example produces a compile error. Are there any examples of the FragmentStore being used like this?

**Comment by martinjvickers** [comment]
@rrahn As for closing the issue, which is a question rather than an issue really, will this mean that the responses from people who might be able to help will dry up as it's a closed issue? 

I ask as it took 2 and a half months for any response so I had obviously had to move onto other work. But not responding for 15 days (which includes the Christmas, New Year period) and having to get back up to speed with what I was trying to do is not unreasonable I think.

I kind of feel that questions like this should be in more of a forum than github issues as once an issue is closed it is effectively hidden as a solved issue. Anyway I'm happy to keep the issue closed as I understand you want to work through what must be a large list of actual issues.

**Comment by rrahn** [comment]
Well, no problem. We can reopen it, sure. Currently, there is no one maintaining it, hence we kept the priority low. But since you are actively interested in this matter, we will take a look into it more closely. 

So after some digging: the`readRecords` function of the fragment store does not support batch wise reading (as @smehringer already informed you about). So the suggested solution was to read your file with `BamFileIn` in batches and add the read alignment information to the store by iteratively using [`appendAlignedRead`](http://docs.seqan.de/seqan/develop/class_FragmentStore.html#FragmentStore%23appendAlignedRead). Make sure you read the contigs before using [`loadContigs`](http://docs.seqan.de/seqan/develop/class_FragmentStore.html#FragmentStore%23loadContigs).
 Of course, this seems a little cumbersome. So adding a batch wise processing should be in general possible. 
I reopen this as a feature request for 2.5 then.

**Comment by martinjvickers** [comment]
Thank you for that, I really appreciate your help, I will give this a go and post my solution when I get a chance.
